### PR TITLE
[Reviewer: Andy] New SAS events

### PR DESCRIPTION
**Pull request by AlexHockey**
This PR contains the new SAS events I am planning on adding to improve the high level events displayed in the detail timeline. 

The main problems we currently have with the detailed timeline are:

* The first event is often in the middle of processing (for a normal call, the first event is the MMTEL TAS looking up simservs). I have added some logs so that the first event explains what was received to kick off processing. 
* Some key checkpoints were missing - for example the start of orig/term/cdiv processing, AS invocations, and the successful completion of a register/subscription request. I have added events for these. 

I am also going to move the Ralf Rf session start/continuation/end events to after Ralf has received the response from the CCF (Ralf already logs each request it receives from Sprout as a level 40 event).

### CODE REVIEW

File: clearwater_sas_resource_bundle.yaml [hunk 1]
```diff
@@ -685,12 +686,6 @@ events:
       Public ID: {{ var_data[0] }}
     level:   80
 
-  0x8100B0:
-    summary: "Get user's simservs from Homer"
-    details: |
-      User: {{ var_data[0] }}
-    level:   80
-
```
Comment (AlexHockey) [inline clearwater_sas_resource_bundle.yaml]: The MMTELTAS always logs this event immediately before `0x860000` so I don't think this is worth keeping.

**Comment by AlexHockey** [comment]
I'll be implementing these events in separate PRs. For now I just want to agree the list of events to add.

**Comment by graemerobertson** [comment]
This looks great, I'm quite excited

File: clearwater_sas_resource_bundle.yaml [hunk 2]
```diff
@@ -750,6 +745,62 @@ events:
...
+  0x8100DF:
+    summary: "Registration request accepted"
+    level: 80
+
+  0x8100E0:
+    summary: "Subscription to registration state accepted"
+    level: 80
+
+  0x8100E1:
+    summary: "Routing request internally to sproutlet {{ var_data[0] }}"
```
Comment (bossmc) [inline clearwater_sas_resource_bundle.yaml]: What about responses?

File: clearwater_sas_resource_bundle.yaml [hunk 3]
```diff
@@ -940,20 +991,42 @@ events:
       IMS subscription: {{ var_data[0] }}
     level:   40
 
+  0x820230:
+    summary: "Received Push Profile Request from the HSS"
```
Comment (bossmc) [inline clearwater_sas_resource_bundle.yaml]: Are these visible in SAS?  How would I ever find them?

Comment (bossmc) [inline clearwater_sas_resource_bundle.yaml]: Public identity as details?

File: clearwater_sas_resource_bundle.yaml [hunk 4]
```diff
@@ -940,20 +991,42 @@ events:
...
+      Received Registration Termination request from the HSS for the IMPI {{ var_data[0] }}, and {{ static_data[0] }} associated IMPI(s).
+    level: 80
+
   #
   # Ralf events (0x830000 - 0x83FFFF)
   #
 
   0x830000:
-    summary: "New Rf session created: {{ var_data[0] }}"
+    summary: "Started Rf session: {{ var_data[0] }}"
```
Comment (bossmc) [inline clearwater_sas_resource_bundle.yaml]: Session IDs are long and random strings should the ID be in the summary?  Or just in the description?

File: clearwater_sas_resource_bundle.yaml [hunk 5]
```diff
@@ -940,20 +991,42 @@ events:
...
+    level:   80
```
Comment (bossmc) [inline clearwater_sas_resource_bundle.yaml]: Do we already have equivalent events for EVENT messages?  ("Sent Rf billing event" or something?)

File: clearwater_sas_resource_bundle.yaml [hunk 6]
```diff
@@ -750,6 +745,62 @@ events:
...
+      Request method: {{ var_data[1] }}
+    level: 80
+
+  0x8100DC:
+    summary: "S-CSCF started originating service processing for {{ var_data[0] }} (following call diversion)"
+    detail: |
+      S-CSCF started orig-cdiv service processing on an initial SIP request.
+      Served user: {{ var_data[0] }}
+      Request method: {{ var_data[1] }}
+    level: 80
```
Comment (bossmc) [inline clearwater_sas_resource_bundle.yaml]: BGCF?  (I think you said these already exist)...

Comment (AlexHockey) [inline clearwater_sas_resource_bundle.yaml]: The BGCF produces a high level log with the output of its routing decision. I'm not sure it is worth adding an explicit even for the BGCF receiving the request.

File: clearwater_sas_resource_bundle.yaml (hunk 3 above)
Comment (graemerobertson) [inline clearwater_sas_resource_bundle.yaml]: When we added SAS logging to homestead, I originally wasn't intending to add logs for PPRs and RTRs, but 'after comments from Alex and Matt' I decided to include them. So there you have it.

Comment (AlexHockey) [inline clearwater_sas_resource_bundle.yaml]: On your first point, they are not currently visible so I was in two minds about adding this event. However Homestead logs other events during PPR processing so I thought I should for consistency. 

On your second point, PPRs contain a full blown IMS subscription, and a private ID (which is used if the PPR is updating charging addresses). Unfortunately: 

*  The IMS subscription isn't parsed until late on in the PPR processing (and I don't want to parse it early just for SAS logging).
*  The IMPI doesn't seem like a sensible thing to log since it isn't always used. 

TL;DR I think I should leave this log as it is.

File: clearwater_sas_resource_bundle.yaml (hunk 4 above)
Comment (AlexHockey) [inline clearwater_sas_resource_bundle.yaml]: Yes I think I agree. I will update.

File: clearwater_sas_resource_bundle.yaml (hunk 6 above)
Comment (bossmc) [inline clearwater_sas_resource_bundle.yaml]: Gemini/Memento?

**Comment by bossmc** [comment]
Reviewed.

File: clearwater_sas_resource_bundle.yaml (hunk 5 above)
Comment (AlexHockey) [inline clearwater_sas_resource_bundle.yaml]: As discussed, I think there is value in having the session-based logs when there are no equivalent event-based logs as:

* I expect people are going to care more about session-based billing than event-based billing. 
* The whole point of Ralf is to handle the stateful nature of session-based billing, so this is the thing most likely to go wrong. 

I'll admit this is just my best guess though.

File: clearwater_sas_resource_bundle.yaml (hunk 2 above)
Comment (AlexHockey) [inline clearwater_sas_resource_bundle.yaml]: Agreed.

File: clearwater_sas_resource_bundle.yaml (hunk 4 above)
Comment (AlexHockey) [inline clearwater_sas_resource_bundle.yaml]: However i *do* think it would be useful to put the role and functionality (e.g. originating S-CSCF) in the summary. Shout if you disagree.

Comment (bossmc) [inline clearwater_sas_resource_bundle.yaml]: Good idea.

File: clearwater_sas_resource_bundle.yaml (hunk 2 above)
Comment (AlexHockey) [inline clearwater_sas_resource_bundle.yaml]: As discussed: I'm a bit worried about the performance impact of raising this event. I am going to remove it for now, and we'll see if the new higher-level events I've added make traces easy enough to debug that we don't need this low-level one.

File: clearwater_sas_resource_bundle.yaml (hunk 6 above)
Comment (AlexHockey) [inline clearwater_sas_resource_bundle.yaml]: Again these (normally) just log a single event explaining what they've done. This could be a problem if we ever deployed them without the sprout S-CSCF but I don't see that happening any time soon.

**Comment by AlexHockey** [comment]
As discussed, I have updated this with:

* Some bug fixes
* Some new events for when the S-CSCF decides it needs to reject a request (the I-CSCF already had the events it needed). 

Please can you re-review.

**Comment by bossmc** [comment]
Your markups look good.
